framework/include/utils/RankFourTensorImplementation.h

```diff
@@ -383,6 +396,33 @@ RankFourTensorTempl<Real>::invSymm() const
   return result;
 }
 
+template <typename T>
+RankFourTensorTempl<T>
+RankFourTensorTempl<T>::inverse() const
+{
+  mooseError("The inverse operation calls to LAPACK and only supports plain Real type tensors.");
+}
+
+template <>
+RankFourTensorTempl<Real>
+RankFourTensorTempl<Real>::inverse() const
+{
+  // The inverse of a 3x3x3x3 in the C_ijkl*A_klmn = de_im de_jn sense is
+  // simply the inverse of the 9x9 matrix of the tensor entries.
+  // So all we need to do is inverse _vals (with the appropriate row-major
+  // storage)
+
+  // Setup 9x9
+  std::vector<PetscScalar> mat(N4);
+  std::copy(_vals, _vals + N4, mat.begin());
+
+  // LAPACK inverse
+  MatrixTools::inverse(mat, N * N);
```

Comment:
1. LAPACK is probably fine for now, but we'd like to move to Eigen, which would support dual numbers. 
2. Can this be done without a copy? Or at least without creating an std::vector? Those are really slow as local variables.

Reply:
One copy is inevitable as we don't want to do the inverse in-place.

We should be able to copy to a new `RankFourTensor` and then invert in place.  But there isn't a `double*` type interface in `MatrixTools` Am I allowed to call LAPACK directly?

Reply:
This can be done elegantly with Eigen using `Eigen::Map<Eigen::Matrix>`. You can map the raw r4t storage to an Eigen object on which you can call invert.

1. Create a R4T return value buffer 
```
RankFourTensorTempl<T> ret;
```

2. Invert the mapped `*this` tensor into the mapped `ret` tensor 
```
Eigen::Map<Eigen::Matrix<Real, 9,9>>(ret._vals, 9, 9) = Eigen::Map<Eigen::Matrix<Real, 9,9>>(_vals, 9, 9).invert();
```

This should be a single copy instruction (most likely the compiler would use RVO to avoid copying altogether).

Reply:
Cool, cool this works.  For my own future reference you need to include `"Eigen/LU"` or else the inverse doesn't ever get defined.